pkg/config/apiserver.go

```diff
@@ -0,0 +1,210 @@
+	pb "istio.io/mixer/pkg/config/proto"
+	"istio.io/mixer/pkg/expr"
+	"istio.io/mixer/pkg/status"
+)
+
+type validateFunc func(cfg map[string]string) (rt *Validated, ce *adapter.ConfigErrors)
+
+type readBodyFunc func(r io.Reader) ([]byte, error)
+
+// API is the server wrapper that listens for incoming requests to the manager and processes them
+type API struct {
+	version  string
+	rootPath string
+
+	// used at the back end for validation and storage
+	store    KVStore
+	validate validateFunc
+
+	// house keeping
+	handler http.Handler
+	Server  *http.Server
+
+	// fault injection
+	readBody readBodyFunc
+}
+
+// DefaultAPIPort default port exposed by the API server.
+const DefaultAPIPort = 9094
+
+// register routes
+func (a *API) register(c *restful.Container) {
+	ws := &restful.WebService{}
+	ws.Consumes(restful.MIME_JSON, "application/yaml", "application/x-yaml")
+	ws.Produces(restful.MIME_JSON)
+	ws.Path(a.rootPath)
+
+	ws.Route(ws.
+		GET("/scopes/{scope}/subjects/{subject}/rules").
+		To(a.getRules).
+		Doc("Gets rules associated with the given scope and subject").
+		Param(ws.PathParameter("scope", "scope").DataType("string")).
+		Param(ws.PathParameter("subject", "subject").DataType("string")).
+		Writes(pb.ServiceConfig{}))
+
+	ws.Route(ws.
+		PUT("/scopes/{scope}/subjects/{subject}/rules").
+		To(a.putRules).
+		Doc("Replaces rules associated with the given scope and subject").
+		Param(ws.PathParameter("scope", "scope").DataType("string")).
+		Param(ws.PathParameter("subject", "subject").DataType("string")).
+		Writes(rpc.Status{}))
+	c.Add(ws)
+}
+
+// NewAPI creates a new API server
+func NewAPI(version string, port int, eval expr.Validator, aspectFinder AspectValidatorFinder,
+	builderFinder BuilderValidatorFinder, findAspects AdapterToAspectMapper, store KVStore) *API {
+	c := restful.NewContainer()
+	a := &API{
+		version:  version,
+		rootPath: fmt.Sprintf("/api/%s", version),
+		store:    store,
+		readBody: ioutil.ReadAll,
+		validate: func(cfg map[string]string) (rt *Validated, ce *adapter.ConfigErrors) {
+			v := newValidator(aspectFinder, builderFinder, findAspects, true, eval)
+			return v.validate(cfg)
+		},
+	}
+	a.register(c)
+	a.Server = &http.Server{Addr: ":" + strconv.Itoa(port), Handler: c}
+	a.handler = c
+	return a
+}
+
+// Run calls listen and serve on the API server
+func (a *API) Run() {
+	glog.Infof("Starting Config API Server at %v", a.Server.Addr)
+	glog.Warning(a.Server.ListenAndServe())
+}
+
+// getRules returns the entire service config document for the scope and subject
+// "/scopes/{scope}/subjects/{subject}/rules"
+func (a *API) getRules(req *restful.Request, resp *restful.Response) {
+	funcPath := req.Request.URL.Path[len(a.rootPath):]
+	val, index, found := a.store.Get(funcPath)
+	if !found {
+		writeResponse(http.StatusNotFound, fmt.Sprintf("no rules for %s\n", funcPath), resp)
+		return
+	}
+
+	// TODO send index back to the client
+	_ = index
+	resp.AddHeader("Content-Type", "application/yaml")
```

Comment:
You already declare content type with `Produces` clause above. Please drop one or the other.